### PR TITLE
Let repo rule attributes reference extension apparent names

### LINKED ISSUE #19055: Provide module extensions with a way to reference extension repos
### What underlying problem are you trying to solve with this feature?

The repository mapping that applies to a module extension does not include the repositories it generates. This makes it very tricky to pass a reference to some file in such a repo to another repo rule instantiated by the same extension, e.g.:

```starlark
def _ext_impl(module_ctx):
    my_lang_toolchain_repo(
        name = "lang_sdk",
    )
    my_toolchain_repo(
        name = "lang_toolchain",
        # This does *not* work as `lang_sdk` is not in the current file's repo mapping.
        sdk_root = "@lang_sdk//:root",
    )
```

This is a very common pattern, but requires major hacks to pull of without additional API: 
1. Observe that the canonical name of an extension repo is `<common prefix>~<apparent name>` and rely on that implementation detail to construct the canonical name of `lang_sdk` in `lang_toolchain` as `ctx.attr.name.removesuffix("lang_toolchain") + "lang_sdk"`. 
2. Make use of the fact that `lang_sdk` is in the repo mapping of the repositories generated by the extension. By making `sdk_root` an `attr.string` instead of an `attr.label` and generating a `.bzl` file in some other ("hub") repo in the the same extension that calls `Label` on a label string, `lang_toolchain` can pass that literal through the correct repo mapping. This requires the module to `use_repo` the hub repo and is overall pretty complex.

### Description of the feature request:

Provide a function on `module_ctx` that converts the name of a repository generated by the extension and a repo-relative label into an instance of `Label` referencing the intended target.

This could look as follows:

```starlark
# In rules_foo v1.2.3, extensions.bzl
def _ext_impl(module_ctx):
    repo(name = "foo", ...)
    repo(name = "bar", ...)

    # Returns `Label("@rules_foo~1.2.3~ext~foo//dir:name")`
    module_ctx.extension_repo_label("@foo//dir:name")
    # Returns `Label("@rules_foo~1.2.3~ext~bar//:name")`
    module_ctx.extension_repo_label("@bar//:name")

    # Fails immediately, argument has to be absolute
    module_ctx.extension_repo_label("//:name")
    # Fails immediately, label must not be canonical
    module_ctx.extension_repo_label("@@rules_foo~1.2.3~ext~foo//dir:name")
    # Fails immediately, repo name must be a valid "user-specified" name
    module_ctx.extension_repo_label("@rules_foo~1.2.3~ext~foo//dir:name")
    # Fails after the module extension returns as `not_foo` is not generated by the extension.
    module_ctx.extension_repo_label("@not_foo//dir:name")
```

Alternative ideas for names: 
* `label_in_repo`

Alternative signatures: 
* Separate arguments for apparent repository name and repo-relative label.

Alternative approaches:
* Hook into the repo rule creation and defer the conversion of label strings into `Label` instances until after the module extension has fully evaluated, at which point the names of all the repos it generates are known. Then use this information to convert label strings implicitly. This has the advantage that the "obvious" approach just works and the downside that its implementation is probably more involved.

### Which operating system are you running Bazel on?

N/A

### What is the output of `bazel info release`?

6.3.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by fmeum**
Fixes #19055

RELNOTES: Repository rules instantiated in the same module extensions can now refer to each other by their extension-specified names in label attributes.

### CODE REVIEW

**Comment by fmeum** [comment]
This is currently stacked on #23371.

**Comment by fmeum** [comment]
I fixed up the error messages and also rebased on master.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionEvalStarlarkThreadContext.java [hunk 1]
```diff
@@ -56,94 +59,164 @@ public static ModuleExtensionEvalStarlarkThreadContext fromOrNull(StarlarkThread
...
-      RepoSpec repoSpec =
-          RepoSpec.builder()
-              .setBzlFile(bzlFile)
-              .setRuleClassName(ruleClass.getName())
-              .setAttributes(attributesValue)
-              .build();
+  /**
+   * Evaluates the repo rule calls recorded by {@link #lazilyCreateRepo} and returns all repos
+   * generated by the extension. The key is the "internal name" (as specified by the extension) of
+   * the repo, and the value is the package containing (only) the repo rule.
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionEvalStarlarkThreadContext.java:125]: the value is the RepoSpec these days. we should've updated the documentation before

**Review by Wyverald — COMMENTED** [review]
overall looks nice!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java [hunk 2]
```diff
@@ -967,11 +970,17 @@ public RunModuleExtensionResult run(
             moduleContext.getRecordedFileInputs(),
             moduleContext.getRecordedDirentsInputs(),
             moduleContext.getRecordedEnvVarInputs(),
-            threadContext.getGeneratedRepoSpecs(),
+            threadContext.createAndGetRepos(starlarkSemantics),
             moduleExtensionMetadata,
             repoMappingRecorder.recordedEntries());
       } catch (EvalException e) {
-        env.getListener().handle(Event.error(e.getMessageWithStack()));
+        Location innermostLocation =
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java]: I added a helper method and updated all other cases I could find. It's slightly duplicative information as the stack trace also contains locations, but this gets rid of the `<builtin>` and is also easier to parse and copy-paste.

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

**Pull request closed without merging**